### PR TITLE
Render user supplied data in UserWidget

### LINKED ISSUE #1945: Missing user supplied data in User interface
We need this.

### PR DESCRIPTION
**Pull request by mattrobenolt**
fixes gh-1945

/cc @benvinegar

### CODE REVIEW

File: src/sentry/static/sentry/app/views/groupDetails/userWidget.jsx [hunk 1]
```diff
@@ -1,6 +1,25 @@
...
+  return key.replace(/_/g, ' ');
+}
+
+function renderLine(key, value, name='') {
+  if (name === '') {
+    name = keyToName(key);
+  }
+  return [
+    <dt key={key + "-label"}>{name}:</dt>,
+    <dd key={key}>{value.toString()}</dd>
```
Comment (benvinegar) [inline src/sentry/static/sentry/app/views/groupDetails/userWidget.jsx]: I am sure that `value` is coerced to a string.

Comment (mattrobenolt) [inline src/sentry/static/sentry/app/views/groupDetails/userWidget.jsx]: It wasn't behaving correctly when value was a boolean, fwiw. Which is why I added this.

Lemme dubble check.

File: src/sentry/static/sentry/app/views/groupDetails/userWidget.jsx [hunk 2]
```diff
@@ -1,6 +1,25 @@
...
 import React from "react";
 import Gravatar from "../../components/gravatar";
 
+function keyToName(key) {
+  // Take a given key, and transform it from
+  // camel case to title case
+  if (!key) return '';
+  key = key[0].toUpperCase() + key.slice(1);
+  return key.replace(/_/g, ' ');
+}
```
Comment (benvinegar) [inline src/sentry/static/sentry/app/views/groupDetails/userWidget.jsx]: Won't this just convert the first character to upper case?

Why not use [this implementation](http://stackoverflow.com/questions/196972/convert-string-to-title-case-with-javascript) instead?

File: src/sentry/static/sentry/app/views/groupDetails/userWidget.jsx (hunk 1 above)
Comment (mattrobenolt) [inline src/sentry/static/sentry/app/views/groupDetails/userWidget.jsx]: Yeah, with a boolean, it renders empty nodes.

File: src/sentry/static/sentry/app/views/groupDetails/userWidget.jsx (hunk 2 above)
Comment (mattrobenolt) [inline src/sentry/static/sentry/app/views/groupDetails/userWidget.jsx]: > Won't this just convert the first character to upper case?

Yeah, that was the intent.

File: src/sentry/static/sentry/app/views/groupDetails/userWidget.jsx [hunk 3]
```diff
@@ -1,6 +1,25 @@
...
+
+function renderLine(key, value, name='') {
+  if (name === '') {
```
Comment (benvinegar) [inline src/sentry/static/sentry/app/views/groupDetails/userWidget.jsx]:
```javascript
name = name || keyToName(key);
```

Comment (mattrobenolt) [inline src/sentry/static/sentry/app/views/groupDetails/userWidget.jsx]: Just did `{name || keyToName(key)}`

File: src/sentry/static/sentry/app/views/groupDetails/userWidget.jsx [hunk 4]
```diff
@@ -8,28 +27,25 @@ var UserWidget = React.createClass({
...
+    // Handle our native attributes special
+    user.id && children.push(renderLine('id', user.id, 'ID'));
+    user.email && children.push(renderLine('email', user.email));
+    user.username && children.push(renderLine('username', user.username));
+    user.ipAddress && children.push(renderLine('ipAddress', user.ipAddress, 'IP'));
+
+    // We also attach user supplied data as 'user.data'
+    _.each(user.data, function(value, key) {
+      children.push(renderLine(key, value));
+    });
```
Comment (benvinegar) [inline src/sentry/static/sentry/app/views/groupDetails/userWidget.jsx]: I'd probably write as:

```javascript
var children = _.chain(user)
  .pick(user, ['id', 'email', 'username', 'ipAddress'])
  .map((val, key) => {
    key = key === 'id' ? 'ID' :
      key === 'ipAddress' ? 'IP' : key;

    return renderLine(key, value);
  })
  .value();
```

Comment (mattrobenolt) [inline src/sentry/static/sentry/app/views/groupDetails/userWidget.jsx]: Do you want me to change it? Or can I leave as it?

Personally, I think my version is more readable. :)

File: src/sentry/static/sentry/app/views/groupDetails/userWidget.jsx (hunk 1 above)
Comment (benvinegar) [inline src/sentry/static/sentry/app/views/groupDetails/userWidget.jsx]: Ah, yeah, I guess React likely ignores false-y values. Nevermind.

File: src/sentry/static/sentry/app/views/groupDetails/userWidget.jsx (hunk 4 above)
Comment (benvinegar) [inline src/sentry/static/sentry/app/views/groupDetails/userWidget.jsx]: You can leave it. The 'ID' / 'IP' thing makes it less pretty. I also left out `user.data`.

Comment (mattrobenolt) [inline src/sentry/static/sentry/app/views/groupDetails/userWidget.jsx]: :+1:

**Comment by benvinegar** [comment]
:+1:

**Pull request merged**